### ISSUE TITLE
[BUG] Safe policy modules that call _project on a smaller test env crash

### ISSUE THREAD

**Issue by matteobettini** [pytorch/rl#1018]
Let's say we have 2 environments:
-  one `env` for collection with action spec with shape `[500,3,2]`,  with [500] batch_size
-  one `test_env` for testing with action spec with shape `[1,3,2]`,  with [1] batch_size

We create a safe policy 
```python
policy_module = ProbabilisticActor(
        module=policy_module,
        spec=env.action_spec, # `[500,3,2]`
        in_keys=["loc", "scale"],
        distribution_class=IndependentNormal,
        safe=True,
    )
```
 and collect with it in the `env`.

Now, if we run `test_env.rollout(policy_module)` we get an error when the policy tries to project its action to the `test_env.action_spec`, because it was created using `env.action_spec`.

What is the cleanest way to solve this usecase?

##  Proposed solution

Make the `project` methods more flexible?
Or is the solution to keep a different policy for training and one for testing with the 2 separate specs?

**Comment by vmoens** [comment]
I usually solve this by passing the spec of a non-batched env, would that work here?

**Comment by matteobettini** [comment]
For the first issue yes, we could use the `test_env.input_spec` also for the batched one. But this is not super clear IMO.

For the second issue, that is not applicble i think

**Comment by vmoens** [comment]
I'm not sure what your feature dimension is, but what I meant is not having a spec of shape `[1, 3, 2]` but more like `[3, 2]` or `[2]` depending on what the features are.
Happy to consider other solutions if that does not fit. We could implement a `__getitem__` in the specs to be able to reduce it easily:
```
policy = Actor(..., spec=my_spec[0, 0])
```

**Comment by matteobettini** [comment]
```python
base_env = GymEnv("InvertedDoublePendulum-v4")
base_env = TransformedEnv(
    base_env,
    Compose(
        # normalize observations

        DoubleToFloat(
            in_keys=["observation"],
        ),

    ))
env = SerialEnv(3, lambda: base_env)
env.start()
env.action_spec.space.minimum[:] = -0.5

actor_net = nn.Sequential(
    nn.LazyLinear(2 * env.action_spec.shape[-1]),
    NormalParamExtractor(),
)
policy_module = TensorDictModule(
    actor_net, in_keys=["observation"], out_keys=["loc", "scale"]
)
policy_module = ProbabilisticActor(
    module=policy_module,
    spec=env.action_spec,
    in_keys=["loc", "scale"],
    distribution_class=TanhNormal,
    distribution_kwargs={
        "min": env.action_spec.space.minimum,
        "max": env.action_spec.space.maximum,
    },
    return_log_prob=True,
    # we'll need the log-prob for the numerator of the importance weights
)

policy_module(env.reset())

td = env.rollout(10)

data_view = td.reshape(-1)

dist = policy_module.get_dist(td)
```

**Comment by matteobettini** [comment]
The penultimate reshape line is the one that causes the bug.

And we need that to have randomness in the replaybuffer

**Comment by vmoens** [comment]
You can do
```python
    distribution_kwargs={
        "min": base_env.action_spec.space.minimum,
        "max": base_env.action_spec.space.maximum,
    },
```
which solves the issue.

> And we need that to have randomness in the replaybuffer

What does that mean?

**Comment by matteobettini** [comment]
> What does that mean?

I thought that if you pass a td of shape [B,T] to the repaly buffer, you are not mixing up trajectories.
So I thought that the reason why we do `data_view = td.reshape(-1)` in the ppo tutorial is to mix trajectories.

> ```python
>     distribution_kwargs={
>         "min": base_env.action_spec.space.minimum,
>         "max": base_env.action_spec.space.maximum,
>     },
> ```

Environments which have a batch_size in their default version like vmas (which is vectorized) cannot do this.
 I would have to do `"min": base_env.action_spec.space.minimum[0]`

**Comment by vmoens** [comment]
I'm not sure I get it all, I would need more details
You have an environment with a batch-size of `[B]`.
If has specs of batch size `[B, *R]`.
This returns tensordicts of batch size `[B, T]` or possibly `[B, T, *R]`.

the issue is that you provide a spec of size `[B, *R]` and then squash the dims in `[B * T * R]`, that is not expected to work.
If you could squash dims only in `[B*T, R]` and pass specs of shape `[*R]` it would be better. In general, I thought our convention was that all batches that go before the time dimension are irrelevant to MARL/VMAS no?

**Comment by matteobettini** [comment]
> This returns tensordicts of batch size `[B, T]` or possibly `[B, T, *R]`

[B,T]

> If you could squash dims only in `[B*T, R]` and pass specs of shape `[*R]` it would be better.

Yep that is what I am doing, the only thing is that if you have a spec of shape [*B,*R] there is not a function in the specs to get *R. You have to get the batch_size *B from outside the spec (e.g. from the env) to then know what *R is.

**Comment by vmoens** [comment]
Yeah I think that here the best thing is to implement a `__getitem__` for the specs such that you can provide the spec with the minimal shape to the module / dist kwargs. Other than that I'm not sure how to deal with this

**Comment by matteobettini** [comment]
Having a spearation in specs between batch_dim and feature_dim is not possible right? Something like distributions in torch

**Comment by vmoens** [comment]
Why not just `__getitem__`? 
This would solve the problem as well without adding an extra level of complexity.
If we do introduce 2 types of shapes (which always confused me with the distributions) we're asking the users to learn more than they need for most use cases. Then we'll eventually need to port that to the envs bc it's the natural thing to do, then someone will ask for that in tensordict too... 

I'm a bit afraid that this will be a considerable amount of work that will never end.

**Comment by matteobettini** [comment]
Yeah I just wanted to know what you thought about it and this makes sense.
Although I would say that tensordict has this a bit already. 
We use batch_size to limit reshping operations and similars to those dimensions. 

 But I agree, let's not add complexity

**Comment by matteobettini** [comment]
update on this:

with spec indexing we would be able to do
```
with no_batch_size():
   env.action_spec
```
using a function like
```
def _get_spec_without_batch(self, spec_name):
    spec = getattr(self, spec_name)
    idx = [0,] * len(self.batch_size)
    return spec[idx]
```

**Comment by matteobettini** [comment]
Spec indexing introduced in #1105

**Comment by vmoens** [comment]
Closed by #1105
